### PR TITLE
Mark Heapster as Deprecated

### PR DESCRIPTION
**Pull request by DirectXMan12**
This marks Heapster as deprecated, adding a deprecation timeline that
culminates in the retirement of the Heapster project for the Kubernetes
1.13 release.

As per the discussion in SIG Instrumentation.

### CODE REVIEW

**Comment by DirectXMan12** [comment]
cc @kubernetes/heapster-maintainers @kubernetes/sig-instrumentation-misc @brancz @piosz

**Comment by DirectXMan12** [comment]
P.S. I've written up the deprecation timeline as a draft.  I'd like it to be similarly aggressive to what's listed, if possible, but I'm open to suggestions.

**Comment by DirectXMan12** [comment]
reviewers/approvers, please take a look:

@acobaugh, @almogbaku, @andyxning, @bluebreezecf, @burmanm, @crassirostris, @emfree, @ezeev, @huangyuqi, @jamtur01, @johanneswuerbach, @jsoriano, @kawych, @Kokan, @loburm, @mcorbin, @miaoyq, @mwielgus, @mwringe, @piosz, @theairkit, @x13n, @yogeswaran

**Review by jsoriano — APPROVED** [review]

**Comment by kawych** [comment]
/cc @piosz 
I don't think we should fully deprecate Heapster, at least not at this point. For instance, Heapster is currently a go-to monitoring agent for Stackdriver.

I suggest to deprecate only Heapster API and work on replacing it with metrics API served by Metrics Server. My understanding of current monitoring architecture for k8s: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md is that Heapster is going to become just another third-party monitoring agent.

**Comment by brancz** [comment]
@kawych can we find a way to have heapster transition from this org to a stackdriver organization? From the Kubernetes (sig-instrumentation) project point of view we agreed that Heapster is to be depcrecated. That doesn't mean stackdriver can't continue to support it, but as the Kubernetes project as a whole it's not a project we want to continue supporting. Can we agree on deprecating it here and having a stackdriver fork and support as necessary for your organization?

**Comment by DirectXMan12** [comment]
> For instance, Heapster is currently a go-to monitoring agent for Stackdriver.

@brancz beat me to the punch here, but I'll follow up with extended thoughts.

Stackdriver using Heapster is all well and good, but, IMO, that's not a good reason to keep Heapster around as a maintained Kubernetes project -- it's a good reason for GCE to keep around a Heapster fork until they can migrate to a different monitoring solution (or really as long as they want).

One of the key observations of the new monitoring plan was that the Heapster model of development (vendor dump) is unsustainable.  Even apart from that, the Heapster codebase needs significant love.  It has serious design issues (ask me about these later, if you're interested), years accumulated cruft with no real policy around what we can remove, and serious usability issues (I'd be happy to expand on these later as well).

Furthermore, non-deprecated Heapster poses problems from an organization perspective.  People continue to request that we add new sinks, and with the current policies, we don't have a good reason to say no.  It's unclear if this is because they believe that Heapster is still the recommended solution for monitoring Kubernetes (it's not), because they want their monitoring solution to reach a larger audience (which is, in-and-of-itself, also a problem -- see below), or because they don't want to maintain collection code.

On that subject, from a Kubernetes perspective, we want people to invest in a healthy monitoring ecosystem with solutions designed to work with their tools from the beginning, as opposed to shoehorning the tools into Heapster's architecture -- we have a number of PRs like "this specific sink has issues with this way the metrics are structured in Heapster, so we need to work around it".  I'm not try to cast a bad light on our sink maintainers -- they do a wonderful job with the resources they have, but it's tough trying to keep things working, between the core Heapster code and the sink-specific hacks.

As an added benefit, this'll help ensure that, as we move forward with features that require use of the new monitoring architecture, we can be more certain that distributions have either switched over to using it, or made a conscious decision to maintain their own copy of Heapster.

To summarize: one of the main reasons for the new monitoring architecture was to remove Heapster as a bottleneck to development, and to ensure more maintainable code.  Keeping Heapster around as  a maintained Kubernetes project hinders that effort.

**Comment by brancz** [comment]
Thanks for taking the time to express yourself @DirectXMan12! You have clearly laid out, what my thoughts are exactly (but expressed them poorly myself).

**Comment by andyxning** [comment]
> Is metrics-server going to be promoted from incubator?
Agreed that we need also promote metrics-server to kubernetes organization alongside the deprecation of Heapster if we do need to deprecate Heapster.

> I suggest to deprecate only Heapster API and work on replacing it with metrics API served by Metrics Server. My understanding of current monitoring architecture for k8s: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md is that Heapster is going to become just another third-party monitoring agent.

It also seems to me that keep the scrape and sink ability for Heapster should be ok. IMHO, moving to metrics api, metrics server is good for HPA to better support custom metrics which is the currently the big pain point for instrumentation. So, i think we should move the api functionality from Heapster to metrics api.

IMHO, Heapster deprecation is too sudden. There are no simple and easy setup guides for metrics api, metrics server and third party tsdb servers pipeline available to end users to learn and try for now which is a big gap between Heapster. Actually i, even as a developer, has no clear thinking about how to set up a monitor system with metrics api, metrics server and third party tsdb to replace the existing one backed by Heapster with no big changes. If we want end users move to the new monitor pipeline, we should add some docs or guides to help them to migrate from Heapster backed monitoring system to the new one.

Also there are some existing add-ons or functionalities depends on the data from Heapster. What is the procedure to help those components migrating to the new pipeline smoothly.

BTW, what is the replacement for `eventer`. People may still be using it.

> one of the main reasons for the new monitoring architecture was to remove Heapster as a bottleneck to development, and to ensure more maintainable code. Keeping Heapster around as a maintained Kubernetes project hinders that effort.
Agreed that maintaining Heapster with multi sinks is painful. But maybe we should think a way for Heapster to support a plugin mechanism just like out-of-tree cloud provider support or different device support with device plugin instead of kicking it off and setup a new one. :)

**Comment by m1093782566** [comment]
It's really a very sudden news to me! I was thinking contributing a new sink([poseidon](https://github.com/kubernetes-sigs/poseidon)) which is leaded by sig-scheduling.

**Comment by deepak-vij** [comment]
Yeah, our current ongoing Poseidon/Firmament scheduler effort leverages Heapster real-time metrics for resource scheduling. We created a new sink for pushing metrics information out of heapster to Firmament scheduler.

As Dujun mentioned, we were about to initiate new Poseidon sink upstreaming process in the next couple of days.

**Comment by andyxning** [comment]
@DirectXMan12 Thanks for the detailed and excellent explanation.

Sorry for not catching up with the latest usage about HPA/kubectl top and Heapster. :)

Maybe we could do this better in the following aspects:
* Add the deprecation summary for Heapster also to the Kubernetes release v1.11 release note and update the deprecation info for Kubernetes release v1.12 and v1.13. Most people will try to read the summary release note for Kubernetes to learn the trends for Kubernetes. So, i think it is also a good place to make the heapster deprecation widely known.
* Add a detailed setup docs for metrics api, metrics server and third party monitoring system to learn.
* Add the replacement binary for `eventer` in case some people still using it.

**Comment by kawych** [comment]
@DirectXMan12 @brancz 
Thank you for detailed explanations. I've also discussed with @piosz the Stackdriver case and once he approves, I have no objections. Removing myself from reviewers.

**Comment by acobaugh** [comment]
As the very short-term co-maintainer of the influxdb sink, I fully support this effort. For the influxdb case, I believe telegraf with its prometheus input plugin ought to be able to replace the functionality of heapster by pulling metrics straight from metrics-server. I haven't dug into this yet, but I wouldn't be surprised if someone in the TICK community is already doing just that, and we just need to document those methods and configurations within the kubernetes project as an example to get metrics into influxdb, and as a migration path for the heapster+influxdb users out there.

**Comment by DirectXMan12** [comment]
> Add the deprecation summary for Heapster also to the Kubernetes release v1.11 release note and update the deprecation info for Kubernetes release v1.12 and v1.13. Most people will try to read the summary release note for Kubernetes to learn the trends for Kubernetes. So, i think it is also a good place to make the heapster deprecation widely known.

Yes, definitely!

> Add a detailed setup docs for metrics api, metrics server and third party monitoring system to learn.

I'd love to improve the docs.  Can you put together a specific list of complaints you have with the official setup docs in the Kubernetes documentation?

> Add the replacement binary for eventer in case some people still using it.

We should definitely link to that somewhere.  Does anybody remember exactly what it was?  I can't recall.  I'll need to track it down otherwise.

> For the influxdb case, I believe telegraf with its prometheus input plugin ought to be able to replace the functionality of heapster by pulling metrics straight from metrics-server.

It shouldn't be pulling from metrics-server.  It should query the nodes instead.  At any rate, we should have follow-up PRs to build up a list of alternatives for the third-party monitoring solutions.

@piosz any objections to this timeline.  Otherwise, I'll plan on merging by the end of the week.

**Comment by DirectXMan12** [comment]
Since there are no further maintainer objections lodged, I'm going to merge this.

**Comment by eherot** [comment]
Hey guys, were any decisions ever made W/R/T the replacement for `eventer`? I see a bunch of discussion of linking to a replacement but with no follow-up...

**Comment by DirectXMan12** [comment]
It differs based on sink.  I believe there was a fluentd event exporter at one point, and there's now a stackdriver event exporter, plus what appears to be a more generic one at https://github.com/alauda/event-exporter (not endorsing that one, but it does appear to be a potential solution).

**Comment by DirectXMan12** [comment]
There's also Heptio's eventrouter, it appears: https://github.com/heptiolabs/eventrouter
